mempool/v1/mempool.go

```diff
@@ -0,0 +1,414 @@
+// when a block proposer constructs a block and a thread-safe linked-list that
+// is used to gossip transactions to peers in a FIFO manner.
+type TxMempool struct {
+	logger       log.Logger
+	metrics      *mempool.Metrics
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+
+	// txsAvailable fires once for each height when the mempool is not empty
+	txsAvailable         chan struct{}
+	notifiedTxsAvailable bool
+
+	// height defines the last block height process during Update()
+	height int64
+
+	// sizeBytes defines the total size of the mempool (sum of all tx bytes)
+	sizeBytes int64
+
+	// cache defines a fixed-size cache of already seen transactions as this
+	// reduces pressure on the proxyApp.
+	cache mempool.TxCache
+
+	// txStore defines the main storage of valid transactions. Indexes are built
+	// on top of this store.
+	txStore *TxStore
+
+	// gossipIndex defines the gossiping index of valid transactions via a
+	// thread-safe linked-list. We also use the gossip index as a cursor for
+	// rechecking transactions already in the mempool.
+	gossipIndex *clist.CList
+
+	// recheckCursor and recheckEnd are used as cursors based on the gossip index
+	// to recheck transactions that are already in the mempool. Iteration is not
+	// thread-safe and transaction may be mutated in serial.
+	recheckCursor *clist.CElement // next expected response
+	recheckEnd    *clist.CElement // re-checking stops here
+
+	// priorityIndex defines the priority index of valid transactions via a
+	// thread-safe priority queue.
+	priorityIndex *TxPriorityQueue
+
+	// A read/write lock is used to safe guard updates, insertions and deletions
+	// from the mempool. A read-lock is implicitly acquired when executing CheckTx,
+	// however, a caller must explicitly grab a write-lock via Lock when updating
+	// the mempool via Update().
+	mtx       tmsync.RWMutex
+	preCheck  mempool.PreCheckFunc
+	postCheck mempool.PostCheckFunc
+}
+
+func NewTxMempool(
+	logger log.Logger,
+	cfg *config.MempoolConfig,
+	proxyAppConn proxy.AppConnMempool,
+	height int64,
+	options ...TxMempoolOption,
+) *TxMempool {
+
+	mp := &TxMempool{
+		logger:        logger,
+		config:        cfg,
+		proxyAppConn:  proxyAppConn,
+		height:        height,
+		cache:         mempool.NopTxCache{},
+		metrics:       mempool.NopMetrics(),
+		txStore:       NewTxStore(),
+		gossipIndex:   clist.New(),
+		priorityIndex: NewTxPriorityQueue(),
+	}
+
+	if cfg.CacheSize > 0 {
+		mp.cache = mempool.NewLRUTxCache(cfg.CacheSize)
+	}
+
+	// TODO:
+	// proxyAppConn.SetResponseCallback(mp.globalCb)
+
+	for _, opt := range options {
+		opt(mp)
+	}
+
+	return mp
+}
+
+// WithPreCheck sets a filter for the mempool to reject a transaction if f(tx)
+// returns an error. This is executed before CheckTx. It only applies to the
+// first created block. After that, Update() overwrites the existing value.
+func WithPreCheck(f mempool.PreCheckFunc) TxMempoolOption {
```

Comment:
is there a way applications can take advantage of this feature, or is it meant for internal use.

Reply:
I know of no such was. Tbh I don't even know what this is but my hunch is it's meant for internal use only.

Reply:
dam, would be cool to set fee checks here from the app to avoid calling checkTX if the fee is below x.

Reply:
CheckTx is already meant to be "cheap" so I wouldn't worry too much about it. Also, you couldn't use precheck because it'd end up being the same thing as CheckTx -- you need the app to actually process the tx first before you can get the priority.